iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/typed_unique_id.hpp

```diff
@@ -0,0 +1,130 @@
+
+/// @brief Sets the RouDi id to an undefined state
+void unsetUniqueRouDiId() noexcept;
+
+/// @brief returns the unique roudi id
+/// @return value of the unique roudi id
+uint16_t getUniqueRouDiId() noexcept;
+} // namespace internal
+
+/// @brief Struct to signal the constructor to create an invalid id
+struct CreateInvalidId_t
+{
+};
+constexpr CreateInvalidId_t CreateInvalidId = CreateInvalidId_t();
+
+/// @brief Unique ID depending on a type. If you would like to assign different
+///         types consistent unique ids use this class. Every types gets its
+///         own distinct set of ids starting with 0. If the types are the same the
+///         ids are the same.
+///
+/// @code
+///     struct MyClass {
+///         // some members;
+///         iox::port::TypedUniqueId<MyClass> id;
+///     };
+///
+///     struct MySecondClass {
+///         // some members;
+///         iox::port::TypedUniqueId<MySecondClass> id;
+///     };
+//
+///     std::vector<MyClass> myClassVector;
+///     std::vector<MyClass> mySecondClassVector;
+///
+///     uint64_t AddClass() {
+///         myClassVector.emplace_back();
+///         return myClassVector.back().id.getID();
+///     }
+///
+///     uint64_t AddSecondClass() {
+///         mySecondClassVector.emplace_back();
+///         // we use the uint64_t conversion
+///         return mySecondClassVector.back().id;
+///     }
+///
+///     void RemoveClass(const uint64_t id) {
+///         auto iter = std::find_if(myClassVector.begin(), myClassVector.end(), [&](MyClass & c){ return c.id == id;});
+///         if ( iter != myClassVector.end() ) myClassVector.erase(iter);
+///     }
+///
+///     uint64_t id = AddClass();
+///     RemoveClass(id);
+///
+///     // it can be that id == id2 since the id is unique per type
+///     uint64_t id2 = AddSecondClass();
+/// @endcode
+/// @param[in] T type for which the unique ids should be generated
+///
+template <typename T>
+class TypedUniqueId : public cxx::NewType<uint64_t,
+                                          cxx::newtype::ProtectedConstructByValueCopy,
+                                          cxx::newtype::Comparable,
+                                          cxx::newtype::Sortable,
+                                          cxx::newtype::Convertable,
+                                          cxx::newtype::CopyConstructable,
+                                          cxx::newtype::MoveConstructable,
+                                          cxx::newtype::CopyAssignable,
+                                          cxx::newtype::MoveAssignable>
+{
+  public:
+    using ThisType::ThisType;
+
+    /// @brief the constructor creates an id which is greater than the
+    ///         previous created id
+    TypedUniqueId() noexcept;
+
+    /// @brief constructor which creates an invalid id
+    TypedUniqueId(CreateInvalidId_t) noexcept;
+
+    bool isValid() const noexcept;
+
+  private:
+    static constexpr uint64_t INVALID_UNIQUE_ID = 0u;
```

Comment:
An explanation which bits are used for the different id parts would have been useful. I.e. highest 16 bits for the Roudi Id part it seems, the lower 48 bit for the specific part generated by the global counter associated with type T.